PytestPART4: SyslogNgConfig updates

mitzkia opened the pull request:
This patch set adds some configuration features, such as:
- log { flags(...); };
- log { log{...}; };
- options{};
- filter {};
New testcases also added for using this features.

kira-syslogng commented:
Build FAILURE

mitzkia commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

furiel commented on tests/pytest_framework/src/syslog_ng_config/renderer.py:
```diff
@@ -39,16 +39,28 @@ def __render(self, re_create_config=None):
+            self.__render_global_options()
         if self.__syslog_ng_config["sources"]:
             self.__render_statements(root_statement="sources", statement_name="source")
+        if self.__syslog_ng_config["filters"]:
```
As discussed IRL, this is not the right direction, and needs a little rethinking. The proposed architecture limits that filters must come in one block, and must happen before any embedded logpath. However, in a valid syslog-ng config, there is no such ordering requirement.
The limitation will be more visible when we will introduce rewrite rules, which would not reorderable with filters in the current architecture.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

mitzkia commented:
@furiel  Thanks for the review comments, I have fixed them. 
Removing WIP flag

mitzkia commented:
I have added 2 refactor commits which make possible to add new configuration features. 
Also added new functional testcases for a new configuration features

mitzkia commented on tests/pytest_framework/src/syslog_ng_config/statement_group.py:
```diff
@@ -26,14 +26,28 @@
-            driver = drivers
-            self.group_node.update(driver.get_driver_node())
+        self.__group_type = group_type
+        self.__group_id = "%s_%s" % (group_type, RandomId(use_static_seed=False).get_unique_id())
```
Ok, agreed. Later we will address this.

mitzkia commented on tests/pytest_framework/functional_tests/logpath/test_flags_catch_all.py:
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python
+#############################################################################
+# Copyright (c) 2015-2018 Balabit
```
Agreed.

mitzkia replied:
will be fixed

mitzkia commented on tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py:
```diff
@@ -47,22 +47,31 @@ def write_config_content(self):
         FileIO(self.__logger_factory, self.__config_path).rewrite(rendered_config)
 
+    @staticmethod
+    def convert_statements_into_list(statements):
```
Agreed, will be fixed

mitzkia commented on tests/pytest_framework/src/syslog_ng_config/statements/logpath/logpath.py:
```diff
@@ -53,3 +58,10 @@ def update_logpath_with_group(self, group):
+        self.flags.append(flag)
+
+    def add_flags(self, flags):
+        for flag in flags:
```
Ok, will be fixed.

mitzkia commented on tests/pytest_framework/functional_tests/logpath/test_flags_catch_all.py:
```diff
@@ -0,0 +1,132 @@
+def test_flags_catch_all(tc):
+    # Check the correct output if the logpath is the following
+    # log {
+    #     source(s_file);
```
Ok, agreed, Test it with such config:
    # log {
    #     source(s_file);
    #     destination(d_file1);
    # };
    # log { destination(d_file4); flags(catch-all);};

mitzkia commented on tests/pytest_framework/functional_tests/logpath/test_flags_fallback.py:
```diff
@@ -0,0 +1,127 @@
+    # Check the correct output if the logpath is the following
+    # log {
+    #     source(s_file);
+    #     log { filter{host("host-A")}; destination(d_file1); };
```
Ok, agreed, let test with simpleer config.

mitzkia commented on tests/pytest_framework/functional_tests/logpath/test_flags_final.py:
```diff
@@ -0,0 +1,127 @@
+    # log {
+    #     source(s_file);
+    #     log { filter{host("host-A")}; destination(d_file1); flags(final); };
+    #     log { filter{program("app-A")}; destination(d_file2); };
```
Ok, agreed, let test with simpler config

mitzkia commented on tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py:
```diff
@@ -47,27 +53,50 @@ def write_config_content(self):
+        return filter_group
 
-    def create_logpath(self, sources, destinations):
+    def create_logpath(self, logpath_groups=None, flags=None, top_level_logpath=True):
```
Ok, agreed.

mitzkia commented on tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py:
```diff
@@ -91,12 +91,12 @@ def create_filter_group(self, filters):
         return filter_group
 
-    def create_logpath(self, logpath_groups=None, flags=None):
+    def create_logpath(self, logpath_groups=None, flags=None, top_level_logpath=True):
```
Ok, I like the add_inner_logpath() way

mitzkia commented:
@furiel Thanks for the review comments, I have added Fixup commits (for better reading). After they are accepted I will squash them.

mitzkia commented on tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py:
```diff
@@ -47,27 +53,50 @@ def write_config_content(self):
+        return filter_group
 
-    def create_logpath(self, sources, destinations):
+    def create_logpath(self, logpath_groups=None, flags=None, top_level_logpath=True):
```
Ok, I am also agreed with this part. As we IRL discussed later we will handle this.

kira-syslogng commented:
Build SUCCESS

furiel commented on tests/pytest_framework/functional_tests/logpath/test_flags_catch_all.py:
```diff
@@ -0,0 +1,76 @@
+    destination_group3 = config.create_destination_group(file_destination3)
+
+    file_destination4 = config.create_file_destination(file_name="output4.log")
+    destination_group4 = config.create_destination_group(file_destination4)
```
Can you update the numbering too? Or I would even prefer using more talkative names: `source_group`, `file_destination`, `catchall_file_destination`.

mitzkia replied:
Thanks, ok. Will be updated.

furiel commented on tests/pytest_framework/functional_tests/logpath/test_flags_catch_all.py:
```diff
@@ -0,0 +1,76 @@
+    config.create_logpath(statements=[destination_group4], flags="catch-all")
+
+    bsd_message1 = write_message_with_fields(
+        tc, file_source, hostname="host-A", program="app-A", message="message from host-A and app-A"
```
Setting hostname, program, and keep-hostname() global options is not needed anymore.

mitzkia replied:
Thanks, ok. Will be updated.

furiel commented on tests/pytest_framework/functional_tests/logpath/test_flags_fallback.py:
```diff
@@ -0,0 +1,91 @@
+    destination_group1 = config.create_destination_group(file_destination1)
+
+    file_destination3 = config.create_file_destination(file_name="output3.log")
+    destination_group3 = config.create_destination_group(file_destination3)
```
Can you fix the numbering here too?

mitzkia replied:
Thanks, ok. Will be updated.

furiel commented on tests/pytest_framework/functional_tests/logpath/test_flags_final.py:
```diff
@@ -0,0 +1,92 @@
+
+    second_level_logpath1 = config.create_inner_logpath(statements=[filter_group1, destination_group1], flags="final")
+
+    second_level_logpath3 = config.create_inner_logpath(statements=[destination_group3])
```
Can you fix the numbering here too?

mitzkia replied:
Thanks, ok. Will be updated.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS
